### PR TITLE
[MC-754][MC-838][MC-839][MC-840] PDF Reader plugin - Fix Extract Text action

### PR DESCRIPTION
**Pull request by dsliwinski-r7**
### Description

Describe the proposed changes:

 - fix Extract Text action

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "rapid7/PDF https://example.com Step name: extract_text\n",
    "meta": {},
    "output": {
      "output": "9/17/21, 6:37 AM TEST PDF https://example.com happy  jaychow  pdf-file  reading  file:///C:/Users/jchow/Desktop/TEST PDF https://example.com 1/1"
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/pdf_reader:1.0.2 --debug run < tests/extract_text.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/PDF https://example.com Step name: extract_text\n",
    "meta": {},
    "output": {
      "output": "Example Title text text textsample text"
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/pdf_reader:1.0.2 --debug run < tests/extract_text2.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/PDF https://example.com Step name: extract_text\n",
    "meta": {},
    "output": {
      "output": "sample text"
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/pdf_reader:1.0.2 --debug run < tests/extract_text3.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nThe provided content is not from the PDF file. Please check that the input is correct and try again.",
    "log": "rapid7/PDF https://example.com Step name: extract_text\nAn error occurred during plugin execution!\n\nThe provided content is not from the PDF file. Please check that the input is correct and try again.\nTraceback (most recent call last):\n  File \"/usr/local/lib/https://example.com\", line 25, in run\n    pdf_file = https://example.com\n  File \"/usr/local/lib/https://example.com\", line 60, in open\n    return cls(path_or_fp, **kwargs)\n  File \"/usr/local/lib/https://example.com\", line 33, in __init__\n    https://example.com = PDFDocument(PDFParser(stream), password=password)\n  File \"/usr/local/lib/https://example.com\", line 572, in __init__\n    raise PDFSyntaxError('No /Root object! - Is this really a PDF?')\https://example.com No /Root object! - Is this really a PDF?\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/https://example.com\", line 326, in handle_step\n    output = https://example.com\n  File \"/usr/local/lib/https://example.com\", line 476, in start_step\n    output = func(params)\n  File \"/usr/local/lib/https://example.com\", line 33, in run\n    raise PluginException(\https://example.com An error occurred during plugin execution!\n\nThe provided content is not from the PDF file. Please check that the input is correct and try again.\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/pdf_reader:1.0.2 --debug run < tests/extract_text_bad.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nThe provided content is not from the PDF file. Please check that the input is correct and try again.",
    "log": "rapid7/PDF https://example.com Step name: extract_text\nAn error occurred during plugin execution!\n\nThe provided content is not from the PDF file. Please check that the input is correct and try again.\nTraceback (most recent call last):\n  File \"/usr/local/lib/https://example.com\", line 25, in run\n    pdf_file = https://example.com\n  File \"/usr/local/lib/https://example.com\", line 60, in open\n    return cls(path_or_fp, **kwargs)\n  File \"/usr/local/lib/https://example.com\", line 33, in __init__\n    https://example.com = PDFDocument(PDFParser(stream), password=password)\n  File \"/usr/local/lib/https://example.com\", line 572, in __init__\n    raise PDFSyntaxError('No /Root object! - Is this really a PDF?')\https://example.com No /Root object! - Is this really a PDF?\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/https://example.com\", line 326, in handle_step\n    output = https://example.com\n  File \"/usr/local/lib/https://example.com\", line 476, in start_step\n    output = func(params)\n  File \"/usr/local/lib/https://example.com\", line 33, in run\n    raise PluginException(\https://example.com An error occurred during plugin execution!\n\nThe provided content is not from the PDF file. Please check that the input is correct and try again.\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/pdf_reader:1.0.2 --debug run < tests/extract_text_non_pdf_file_bad.json
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin failed validation! The following validation errors occurred:
Validator "SupportedVersionValidator" failed! 
	Cause: supported_versions is missing from plugin.spec.yaml.

----
[*] Total time elapsed: 8414.214ms
```

<summary>
icon-validate --all .
</summary>
</details>

### CODE REVIEW

File: plugins/pdf_reader/komand_pdf_reader/actions/extract_text/action.py [hunk 1]
```diff
@@ -1,41 +1,37 @@
...
+                pdf_file = pdfplumber.open(f)
+                try:
+                    pages = pdf_file.pages
+                    for page in enumerate(pages):
+                        pdf_text += page[1].extract_text().replace("\n", " ")
+                finally:
+                    pdf_file.close()
+        except PDFSyntaxError:
+            raise PluginException(
+                cause="The provided content is not from the PDF file.",
```
Comment (pmara-r7) [inline plugins/pdf_reader/komand_pdf_reader/actions/extract_text/action.py:34]:
```suggestion
                cause="The provided content is not from the PDF file.",
```
```suggestion
                cause="The provided content is not in PDF file format.",
```

**Review by pmara-r7 — CHANGES_REQUESTED** [review]

File: plugins/pdf_reader/plugin.spec.yaml [hunk 2]
```diff
@@ -29,6 +29,7 @@ actions:
...
...Uy9EZXZpY2VSR0IvSSB0cnVlPj4vQ29udGVudHMgMiAwIFI+PgplbmRvYmoKCjQgMCBvYmoKPDwvVHlwZS9QYWdlcwovUmVzb3VyY2VzIDExIDAgUgovTWVkaWFCb3hbIDAgMCA1OTUgODQxIF0KL0tpZHNbIDEgMCBSIF0KL0NvdW50IDE+PgplbmRvYmoKCjEyIDAgb2JqCjw8L1R5cGUvQ2F0YWxvZy9QYWdlcyA0IDAgUgovT3BlbkFjdGlvblsxIDAgUiAvWFlaIG51bGwgbnVsbCAwXQovTGFuZyhwbC1QTCkKPj4KZW5kb2JqCgoxMyAwIG9iago8PC9DcmVhdG9yPEZFRkYwMDU3MDA3MjAwNjkwMDc0MDA2NTAwNzI+Ci9Qcm9kdWNlcjxGRUZGMDA0QzAwNjkwMDYyMDA3MjAwNjUwMDRGMDA2NjAwNjYwMDY5MDA2MzAwNjUwMDIwMDAzNjAwMkUwMDM0PgovQ3JlYXRpb25EYXRlKEQ6MjAyMTExMjgyMjQ5NDIrMDEnMDAnKT4+CmVuZG9iagoKeHJlZgowIDE0CjAwMDAwMDAwMDAgNjU1MzUgZiAKMDAwMDAwNjAwOSAwMDAwMCBuIAowMDAwMDAwMDE5IDAwMDAwIG4gCjAwMDAwMDAxOTcgMDAwMDAgbiAKMDAwMDAwNjE3OCAwMDAwMCBuIAowMDAwMDAwMjE3IDAwMDAwIG4gCjAwMDAwMDUyMDMgMDAwMDAgbiAKMDAwMDAwNTIyNCAwMDAwMCBuIAowMDAwMDA1NDE5IDAwMDAwIG4gCjAwMDAwMDU3MzkgMDAwMDAgbiAKMDAwMDAwNTkyMiAwMDAwMCBuIAowMDAwMDA1OTU0IDAwMDAwIG4gCjAwMDAwMDYyNzcgMDAwMDAgbiAKMDAwMDAwNjM3NCAwMDAwMCBuIAp0cmFpbGVyCjw8L1NpemUgMTQvUm9vdCAxMiAwIFIKL0luZm8gMTMgMCBSCi9JRCBbIDw2RDVGRjZBMERFNDcxNUEyRUZCOURGN0E3MjNFRUFFQT4KPDZENUZGNkEwREU0NzE1QTJFRkI5REY3QTcyM0VFQUVBPiBdCi9Eb2NDaGVja3N1bSAvNDkzMUY3MDAzRjc4RDhBMERENzk2RjlDMTE5RTlDOEQKPj4Kc3RhcnR4cmVmCjY1NDkKJSVFT0YK
```
Comment (cmcnally-r7) [inline plugins/pdf_reader/plugin.spec.yaml:32]: Could use a shorter example if this appears in any UIs
Something like `this is a 9376 length value`

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]

File: plugins/pdf_reader/plugin.spec.yaml (hunk 2 above)
Comment (dsliwinski-r7) [reply plugins/pdf_reader/plugin.spec.yaml:32]: @cmcnally-r7 I can't shorten it any more as it's the encoded whole PDF, not just its contents.

**Review by pmara-r7 — APPROVED** [review]

**Review by mberezin-r7 — DISMISSED** [review]

File: plugins/pdf_reader/help.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 PDF Reader is a plugin for extracting text from a PDF file.
 
-This plugin utilizes a python package called [PyPDF2](https://pypi.org/project/PyPDF2/).
+This plugin utilizes a python package called [pdfplumber](https://pypi.org/project/pdfplumber/).
```
Comment (mrinehart-r7) [inline plugins/pdf_reader/help.md:5]:
```suggestion
This plugin utilizes a Python package called [pdfplumber](https://pypi.org/project/pdfplumber/).
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Review by mrinehart-r7 — APPROVED** [review]

**Pull request merged**